### PR TITLE
Add communication between simulator and native firmware

### PR DESCRIPTION
**Pull request by ecourtois**
Use a shared memory array containing sensors values provided by the simulator.
Add actions to save/load obstacles.
Add a ghost robot to visualize pose order.

### CODE REVIEW

File: cogip/__main__.py [hunk 1]
```diff
@@ -104,7 +104,8 @@ def main():
     # Create robot entity
     robot_entity = RobotEntity(
         asset_path=Path(settings.robot_filename).resolve(),
-        asset_name="Robot2019_Simu"
+        # asset_name="Robot2019_Simu"
```
Comment (gdoffe) [inline cogip/__main__.py:116]: Please remove the commented line

**Review by gdoffe — CHANGES_REQUESTED** [review]
Display: use real robot model.

Use real robot model instead **of** the simplified model
since Qt3D is fast enough to display it instantly.

**Review by gdoffe — APPROVED** [review]

**Review by gdoffe — CHANGES_REQUESTED** [review]
Please can you explain in one sentence what is a layer in the Qt3D.
Because I do not understand the purpose of this commit.

File: cogip/robotentity.py [hunk 2]
```diff
@@ -38,45 +38,57 @@ def post_init(self):
...
+            # {
+            #     "name": "Right sensor",
+            #     "origin_x": -177,
+            #     "origin_y": 0
+            # },
+            # {
+            #     "name": "Left sensor",
+            #     "origin_x": 177,
+            #     "origin_y": 0
+            # },
```
Comment (gdoffe) [inline cogip/robotentity.py:102]: please clean dead code in comments

**Review by gdoffe — CHANGES_REQUESTED** [review]

File: cogip/robotentity.py [hunk 3]
```diff
@@ -135,8 +135,8 @@ def set_dyn_obstacles(self, dyn_obstacles: DynObstacleList) -> None:
             p0, p1, p2, p3 = dyn_obstacle.__root__
             length = math.dist((p0.x, p0.y), (p1.x, p1.y))
             width = math.dist((p1.x, p1.y), (p2.x, p2.y))
-            pos_x = min(p0.x, p1.x) + length/2
-            pos_y = min(p0.y, p2.y) + width/2
+            pos_x = p0.x + (p2.x - p0.x)/2
+            pos_y = p0.y + (p2.y - p0.y)/2
```
Comment (gdoffe) [inline cogip/robotentity.py:150]: Can you add a comment about what are pos_x and pos_y ?

**Review by gdoffe — CHANGES_REQUESTED** [review]
You mean that an already created entity is reused later ?
Can you write a short description of this in your commit message please ?

**Review by gdoffe — CHANGES_REQUESTED** [review]
Please give more context in your commit messages for ignorant guys like me.
What are mesh or cube mesh ?

File: cogip/serialcontroller.py [hunk 4]
```diff
@@ -164,25 +164,35 @@ def get_pose(self):
...
+                self.signal_new_console_text.emit(line)
+                # print(f"parse failed: {line}")
 
     def get_dyn_obstacles(self):
         self.serial_port.write(b"_dyn_obstacles\n")
 
         obstacles_found = False
-        while not self.exiting and not obstacles_found:
+        attempt = 0
+        while not self.exiting and not obstacles_found and attempt < 20:
```
Comment (gdoffe) [inline cogip/serialcontroller.py:200]: it would be good to at least set this treshold into a constant.

**Review by gdoffe — CHANGES_REQUESTED** [review]

**Review by gdoffe — CHANGES_REQUESTED** [review]
Obstacles: setup obstacles size

Increase max length/height and set step to 10 in spinbox.

**Review by gdoffe — CHANGES_REQUESTED** [review]
**Update submodule mcu-firmware

Rebase on master branch including vl53l0x ToF sensor
emulation.**

File: cogip/__main__.py (hunk 1 above)
Comment (ecourtois) [reply cogip/__main__.py:116]: Done

File: cogip/robotentity.py (hunk 2 above)
Comment (ecourtois) [reply cogip/robotentity.py:102]: Done

File: cogip/robotentity.py (hunk 3 above)
Comment (ecourtois) [reply cogip/robotentity.py:150]: pos_x and pos_y are the position of the obstacle, does it really deserve a comment?

File: cogip/serialcontroller.py (hunk 4 above)
Comment (ecourtois) [reply cogip/serialcontroller.py:200]: Done

File: cogip/robotentity.py (hunk 3 above)
Comment (ecourtois) [reply cogip/robotentity.py:130]: Done

**Review by gdoffe — APPROVED** [review]

**Pull request merged**